import random

shell.executable("/bin/bash")
configfile: "analysis.yaml"

__author__ = ['Arnold Kuzniar', 'Luca Santuari']
__version__ = '0.1.0'
__license__ = 'Apache License, Version 2.0'


#ruleorder: samtools_view > bwa_mem
localrules: all

rule all:
    input:
        ### DEBUG:
        #os.path.join(os.path.dirname(config['input']['fasta']), "seqids.fasta")
        #expand(os.path.join("{basedir}", "{genotype}.fasta"),
        #basedir=os.path.dirname(config['input']['fasta']),
        #genotype=config['output']['genotype'])
        #[expand("{prefix}.{ext}", prefix=config['input']['fasta'], ext=['bwt', 'amb', 'ann', 'pac', 'sa']),
        #config['output'].values(),
        # [expand(os.path.join("{basedir}", "{genotype}_{paired}.fq"),
        # basedir=os.path.dirname(config['input']['fasta']),
        # genotype=config['output']['genotype'],
        # paired=[1,2]),
        # expand(os.path.join("{basedir}", "{genotype}{ext}"),
        # basedir=os.path.dirname(config['input']['fasta']),
        # genotype=config['output']['genotype'],
        # ext=config['file_exts']['fasta'][0])]
        expand(os.path.join("{basedir}", "cov{cov}", "{genotype}.bam"),
        basedir=os.path.dirname(config['input']['fasta']),
        cov=config['sim_reads']['coverage'],
        genotype=config['output']['genotype'])

# include rules
include: "rules/sim_genomes.smk"
include: "rules/sim_reads.smk"
include: "rules/map_reads.smk"
