from helper_functions import *

shell.executable('/bin/bash')
configfile: 'analysis.yaml'

__author__ = ['Arnold Kuzniar', 'Luca Santuari']
__version__ = '0.1.0'
__license__ = 'Apache License, Version 2.0'


localrules: all

rule all:
    input:
        ### Used for debugging:
        # os.path.join(config['output']['basedir'], 'seqids.fasta')
        #
        # expand(os.path.join('{basedir}', '{svtype}', '{genotype}.fasta'),
        #        basedir=config['output']['basedir'],
        #        svtype=get_svtype(),
        #        genotype=config['output']['genotype'])
        #
        # expand(os.path.join('{basedir}', '{svtype}', 'r{read_len}_i{insert_len}',
        #                     '{genotype}_{paired}.fq'),
        #        basedir=config['output']['basedir'],
        #        svtype=get_svtype(),
        #        read_len=config['sim_reads']['read']['len'],
        #        insert_len=config['sim_reads']['insert']['len'],
        #        genotype=config['output']['genotype'],
        #        paired=[1,2])
        #
        expand(os.path.join('{basedir}', '{svtype}', 'r{read_len}_i{insert_len}',
                            'cov{cov}', '{genotype}' + get_filext('bam_idx')),
               basedir=config['output']['basedir'],
               svtype=get_svtype(),
               read_len=config['sim_reads']['read']['len'],
               insert_len=config['sim_reads']['insert']['len'],
               cov=config['sim_reads']['coverage'],
               genotype=config['output']['genotype'])

# include rules
include: 'rules/sim_genomes.smk'
include: 'rules/sim_reads.smk'
include: 'rules/map_reads.smk'
