import random

shell.executable("/bin/bash")
configfile: "analysis.yaml"

__author__ = ['Luca Santuari', 'Arnold Kuzniar']
__version__ = '0.1.0'
__license__ = 'Apache License, Version 2.0'


# REFERENCE = config['input']
# NORMALFASTA = config['output']['diploid']
# SVFASTA = config['output']['diploid_sv']
# SEED = random.randint(1, 10**3)

localrules: all

rule all:
    input:
        #os.path.splitext(config['input']['region'])[0] + ".fasta"
        # expand("{basedir}/{genotype}/{prefix}.fasta",
        # basedir=os.path.dirname(config['input']['region']),
        # genotype=config['output']['genotype'],
        # prefix=os.path.splitext(os.path.basename(config['input']['region']))[0])
        #[expand("{prefix}.{ext}", prefix=config['input']['fasta'], ext=['bwt', 'amb', 'ann', 'pac', 'sa']),
        #config['output'].values(),
        [expand(os.path.join("{basedir}", "{genotype}_{paired}.fq"),
        basedir=os.path.dirname(config['input']['region']),
        genotype=config['output']['genotype'],
        paired=[1,2]),
        expand(os.path.join("{basedir}", "{genotype}.sam"),
        basedir=os.path.dirname(config['input']['region']),
        genotype=config['output']['genotype'])]

# include rules
include: "rules/sim_genomes.smk"
include: "rules/sim_reads.smk"
include: "rules/map_reads.smk"
